manpages/bckdatadaily.1
.\"Text automatically generated by txt2man
.TH bckdatadaily 1 "12 May 2014" "" "Backup Scripts Manual"
.SH NAME
\fBbckdatadaily \fP- perform a backup of the files named in backup.files.
.SH SYNOPSIS
.nf
.fam C
\fBbckdatadaily\fP [\fIOPTION\fP]
.fam T
.fi
.fam T
.fi
.SH DESCRIPTION
Perform a backup based on files named in backup.files to a backup file on the
network attached backup share.
.SH OPTIONS
.TP
.B
\fB-h\fP
prints usage information
.TP
.B
\fB-V\fP
prints version information
.SH EXIT STATUS
0
if OK,
.PP
1
if minor problems,
.PP
64
error in command line arguments.
.PP
65
failed to mount share due to mount error.
.PP
66
Backup share already mounted.
.SH AUTHOR
Written by Mark Grant.
.SH REPORTING BUGS
Report bugs to <m.grant.prg@btinternet.com>
.SH COPYRIGHT
Copyright 2010 Mark Grant.
.br
License GPLv3+: GNU GPL version 3 or later
.br
<http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.
.SH SEE ALSO
\fBattbckshare\fP(1) \fBchkbckshare\fP(1) \fBdetbckshare\fP(1) \fBmount\fP(8) (mount_smbfs(8) on FreeBSD) \fBumount\fP(8)


NEWS
Release 1.1.1 2014-01-20
Changed stdout & stderr message output to use a function directing to one or
other based on a status. Changed version option to -V. Added usage of
standard error routine. Introduced function to repeat commands to overcome
latency issues on NAS mount from either mount_smbfs or slow hardware.

Release 1.0.27 2013-10-30
Removed '-' prepending operations in the tar command. Added --exclude-caches
to the tar command. Also changed backup command in bckfulladhoc to use tar or
gtar depending on OS.

Release 1.0.26 2013-09-14
Logic error introduced in bckfullweekly v1.0.19 corrected.

Release 1.0.25 2013-09-14
Improved coding around sequence calculation.

Release 1.0.24 2013-09-14
Changed sequence calculation from DOM / 7 + 1 to testing day
against 7, 14, 21 and 28 as original calc did not work correctly when
day was exactly divisible by 7.

Release 1.0.23 2013-08-22
Debian best practice documentation requires that programs placed in
the PATH should not have extensions as the programs may be rewritten
in a different language. This would then require a name change, or,
if left as-is would cause confusion. So all names changed to remove .sh
extension except for setup.sh which is not on the PATH.

Release 1.0.22 2013-04-01
Changed all scripts to use new backups sub-directory of sysconfdir.Added
include and exclude from files.

Release 1.0.21 2013-03-02
Created new script to support backups from a stored list of files to backup.

Release 1.0.20 2013-02-26
Changed internals of processing command line options.

Release 1.0.19 2013-02-06
Added mailing of backup file date hierachy after backup.

Release 1.0.18 2012-12-20
Added Host name and backup file indicator to email message subject line.

Release 1.0.17 2012-11-17
Changed logic to give responsibility to daily backup to delete the previous
version of that days backup so we now have a full week of full + incremental
backups.

Release 1.0.16 2012-11-06
Reorganised manual pages directory structure in source.

Release 1.0.15 2012-11-06
Updated Author contact information. Moved installation from bindir to sbindir.

Release 1.0.14 2012-11-06
Reverted to use the .sh file extension. Added exclusion to tar command for /run
and /var/run following inclusion of /run in Linux.

Release 1.0.13 2012-01-10
Removed the .sh extension from the command name. Add .gvfs file exclusion to
support Gnome desktops and Ubuntu.

Release 1.0.12 2010-12-16
Allowed mailing of disk usage stats from all OS's

Release 1.0.11 2010-12-14
Modification to an invalid mail switch.

Release 1.0.10 2010-11-28
Modifications to produce and use configuration files for accessing the NAS.

Release 1.0.9 2010-11-23
Daily differential backups for Linux have been restored following a tar bug
fix.

Release 1.0.8 2010-11-21
Re-vamped all manual pages.

Release 1.0.7 2010-11-20
Added manual pages for all commands.

Release 1.0.6 2010-11-20
The daily and weekly scripts no longer call shutdown on cmpletion. Shutdown
must be invoked manually.

Release 1.0.5 2010-11-18
First release of the backup utility scripts.


scripts/bckdatadaily.sh
#! /usr/bin/env bash
##########################################################################
##									##
##	bckdatadaily.sh is automatically generated,			##
##		please do not modify!					##
##									##
##########################################################################

##########################################################################
##									##
## Script ID: bckdatadaily.sh						##
## Author: Mark Grant							##
##									##
## Purpose:								##
## Backs up directories and files specified in backup.files file.	##
## The timing element is expected to be delivered via cron.		##
##									##
## Syntax:	bckincdaily.sh [-h || --help || -v || --version]	##
##									##
## Exit Codes:	0 & 64 - 113 as per C/C++ standard			##
##		0 - success						##
##		64 - Invalid arguments					##
##		65 - Failed mounting backup NAS server			##
##		66 - backup.snar non-existent or inaccesssible		##
##		67 - trap received					##
##									##
##########################################################################

##########################################################################
##									##
## Changelog								##
##									##
## Date		Author	Version	Description				##
##									##
## 02/03/2013	MG	1.0.1	Created.				##
## 01/04/2013	MG	1.0.2	Moved config files to new backups	##
##				directory under sysconfdir. Added	##
##				exclude files for system, weekly, daily	##
##				and data.				##
##									##
##########################################################################

exec 6>&1 7>&2 # Immediately make copies of stdout & stderr

####################
## Init variables ##
####################
script_exit_code=0
version="1.0.2"				# set version variable
etclocation=/usr/local/etc/backups	# Path to etc directory

# Get system name for implementing OS differences
osname=$(uname -s)

###############
## Functions ##
###############
script_short_desc=$(uname -n)" "$(date +%A)" Daily Data Backup"

# Standard function to log $1 to stderr and mail to recipient
mess_log()
{
echo $1 1>&2
echo $1 | mailx -s "$script_short_desc" $mail_recipient
}

# Standard function to log and mail $1, cleanup and return exit code
script_exit()
{
mess_log "$1" 
exec 1>&6 2>&7 6>&- 7>&- # Restore stdin & stdout & close fd's 6 & 7
sleep 5
detbckshare.sh
exit $script_exit_code
}

# Standard trap exit function
trap_exit()
{
script_exit_code=67
script_exit "Script terminating due to trap received. Code: "$script_exit_code
}

# Setup trap
trap trap_exit SIGHUP SIGINT SIGTERM

##########
## Main ##
##########
# Process command line arguments with getopts.
while getopts :hv arg
do
	case $arg in
		h)	echo "Usage is $0 [options]"
			echo "	-h displays usage information"
			echo "	OR"
			echo "	-v displays version information"
			;;
		v)	echo "$0 version "$version
			;;
		\?)	echo "Invalid argument -$OPTARG." >&2
			exit 64
			;;
	esac
done

# If help or version requested then exit now.
if [ $# -gt 0 ]
	then
		exit 0
fi

# Read parameters from $etclocation/backups.conf
IFS="="

exec 3<$etclocation/backups.conf
while read -u3 -ra input
do
	case ${input[0]} in
	dir)
		bckupdir=${input[1]}
		;;
	notifyuser)
		mail_recipient=${input[1]}
		;;
	esac
done
exec 3<&-

# Build the backup file name and path
backpath="/mnt/$bckupdir/backupdata"$(date +%a)".tar.gz"

# Check to see if the NAS backup server is mounted, if not, mount
attbckshare.sh
status=$?
if [ $status != "0" -a $status != "66" ]
	then
	script_exit_code=65
	script_exit "Failed to mount backup NAS server. Mount error: "$status" Script exit code: "$script_exit_code
fi

# Re-direct stdout & stderr to backup logs and write initial entries
exec 1>> /mnt/$bckupdir/backup.log 2>> /mnt/$bckupdir/backuperror.log
echo "Attempting to process backup - "$backpath
mess_log "Attempting to process backup - "$backpath

date
date 1>&2

# If the backup file exists, delete.
# (Just in case full backup has not done this)
if [ -f $backpath -a -r $backpath -a -w $backpath ]
	then
		rm $backpath
fi

# Empty the NAS trashbox
rm /mnt/$bckupdir/trashbox/*

# Get list of sockets to exclude
find / -type s > /mnt/$bckupdir/sockets.exclude

# Run the backup excluding system directories
case $osname in
FreeBSD)
	gtar -cpzf $backpath --files-from=$etclocation/bckdatadaily.files \
		--exclude-from=$etclocation/bcksystem.exclude \
		--exclude-from=$etclocation/bckdatadaily.exclude \
		--exclude-from=/mnt/$bckupdir/sockets.exclude /
	status=$?
;;
Linux)
	tar -cpzf $backpath --files-from=$etclocation/bckdatadaily.files \
		--exclude-from=$etclocation/bcksystem.exclude \
		--exclude-from=$etclocation/bckdatadaily.exclude \
		--exclude-from=/mnt/$bckupdir/sockets.exclude /
	status=$?
;;
esac

# Final log entries and restore stdout & stderr
date
date 1>&2
echo "Processing of "$backpath" is complete. Status: "$status
mess_log "Processing of "$backpath" is complete. Status: "$status
##################################################################
## It is not clear why the following two lines cause everything ##
## to hang until charybdis is restarted. It is probably due to  ##
## the odd wiring relating to the kvm which powers off if	##
## charybdis or scylla are not powered up. So this only affects	##
## priam during EOD backup routines which shutdown machines on	##
## completion. So until fixed test on OS.			##
##################################################################
if [ $osname == "FreeBSD" ]
  then
  df -ah # Log disk stats
  ls -lht /mnt/$bckupdir
fi

# Mail disk stats
df -ah | mailx -s "$script_short_desc" $mail_recipient

# Mail backup file date hierarchy
ls -lht /mnt/$bckupdir | mailx -s "$script_short_desc" $mail_recipient

exec 1>&6 2>&7 6>&- 7>&- # Restore stdout & stderr & close fd's 6 & 7

# Cleanup logs so they only have 1000 lines max
tail -n 1000 /mnt/$bckupdir/backup.log > /mnt/$bckupdir/tmp.log
sleep 5
rm /mnt/$bckupdir/backup.log
sleep 5
mv /mnt/$bckupdir/tmp.log /mnt/$bckupdir/backup.log
sleep 5
tail -n 1000 /mnt/$bckupdir/backuperror.log > /mnt/$bckupdir/tmp.log
sleep 5
rm /mnt/$bckupdir/backuperror.log
sleep 5
mv /mnt/$bckupdir/tmp.log /mnt/$bckupdir/backuperror.log
sleep 5

# Unmount the backup NAS server
detbckshare.sh

# And exit
exit 0
